### ISSUE TITLE
Error output is unclear

### ISSUE THREAD

**Issue by jschewebbn** [Protelis/Protelis#271]
I'm working on upgrading from Protelis 12 to Protelis 13.

I've received similar errors to the one below, however the statement after "In: " was something that I was able to find in my protelis code and fix by adding a final statement of "0" if the last statement in a protelis function was calling a Java void function.

However in this case I'm just getting "In: true" and can't find anything obvious in the protelis code that matches a java function that returns void as the last statement. It seems that this particular case of the error reporting isn't getting enough context to specify where the error is.

I'm executing share_region_nodes.pt from https://github.com/map-dcomp/map-code/blob/bff58f7d988a874d6832c981d5b08db554475aec/src/MAP-Agent/src/test/resources/protelis/com/bbn/map/ap/share_region_nodes.pt and getting the error below.

```
    Caused by: java.lang.NullPointerException: All does not allow null return values. In: true
        at java.util.Objects.requireNonNull(Objects.java:290) ~[?:1.8.0_222]
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.setAnnotation(AbstractAnnotatedTree.java:265) ~[protelis-interpreter-13.0.3.jar:?]
        at org.protelis.lang.interpreter.impl.ShareCall.lambda$evaluate$6(ShareCall.java:134) ~[protelis-interpreter-13.0.3.jar:?]
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.forEachWithIndex(AbstractAnnotatedTree.java:168) ~[protelis-interpreter-13.0.3.jar:?]
        at org.protelis.lang.interpreter.impl.ShareCall.evaluate(ShareCall.java:131) ~[protelis-interpreter-13.0.3.jar:?]
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:123) ~[protelis-interpreter-13.0.3.jar:?]
```

**Comment by DanySK** [comment]
Woah, that's bad.
I believe this is strictly related to #257, are you logging the exceptions via SLF4J?
If so, I expect #270 to fix.
If not, and you are just plainly printing stacktraces, it might be worse.

**Comment by jschewebbn** [comment]
Yes, I'm logging with SLF4J using the log4j backend. This is running inside of junit. I'm not sure what is truncating the stack trace.

**Comment by DanySK** [comment]
...and I imagine you are using log4j as SLF4J backend. I believe I got a clear idea of what's causing the issue then, a PR is on its way

**Comment by jschewebbn** [comment]
There is more to the trace above the "caused by". Perhaps this helps you.
```
    org.protelis.lang.interpreter.util.ProtelisRuntimeException: All does not allow null return values. In: true                                             
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:129) ~[protelis-interpreter-13.0.3.jar:?]                             
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.evalInNewStackFrame(AbstractAnnotatedTree.java:293) ~[protelis-interpreter-13.0.3.jar:?]
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.evalInNewStackFrame(AbstractAnnotatedTree.java:288) ~[protelis-interpreter-13.0.3.jar:?]              
        at org.protelis.lang.interpreter.impl.Invoke.evaluate(Invoke.java:77) ~[protelis-interpreter-13.0.3.jar:?]
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:123) ~[protelis-interpreter-13.0.3.jar:?]                             
        at org.protelis.lang.interpreter.impl.FunctionCall.evaluate(FunctionCall.java:94) ~[protelis-interpreter-13.0.3.jar:?]
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:123) ~[protelis-interpreter-13.0.3.jar:?]                             
        at org.protelis.lang.interpreter.impl.Invoke.evaluate(Invoke.java:95) ~[protelis-interpreter-13.0.3.jar:?]
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:123) ~[protelis-interpreter-13.0.3.jar:?]                             
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.evalInNewStackFrame(AbstractAnnotatedTree.java:293) ~[protelis-interpreter-13.0.3.jar:?]
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.lambda$projectAndEval$2(AbstractAnnotatedTree.java:248) ~[protelis-interpreter-13.0.3.jar:?]          
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.forEachWithIndex(AbstractAnnotatedTree.java:168) ~[protelis-interpreter-13.0.3.jar:?]
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.projectAndEval(AbstractAnnotatedTree.java:247) ~[protelis-interpreter-13.0.3.jar:?]                   
        at org.protelis.lang.interpreter.impl.AssignmentOp.evaluate(AssignmentOp.java:45) ~[protelis-interpreter-13.0.3.jar:?]                                           
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:123) ~[protelis-interpreter-13.0.3.jar:?]                             
        at org.protelis.lang.interpreter.impl.All.lambda$evaluate$0(All.java:52) ~[protelis-interpreter-13.0.3.jar:?]                                                    
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.forEachWithIndex(AbstractAnnotatedTree.java:168) ~[protelis-interpreter-13.0.3.jar:?]                 
        at org.protelis.lang.interpreter.impl.All.evaluate(All.java:50) ~[protelis-interpreter-13.0.3.jar:?]
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:123) ~[protelis-interpreter-13.0.3.jar:?]                             
        at org.protelis.lang.interpreter.impl.FunctionCall.evaluate(FunctionCall.java:94) ~[protelis-interpreter-13.0.3.jar:?]
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:123) ~[protelis-interpreter-13.0.3.jar:?]                             
        at org.protelis.lang.interpreter.impl.Invoke.evaluate(Invoke.java:95) ~[protelis-interpreter-13.0.3.jar:?]                    
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:123) ~[protelis-interpreter-13.0.3.jar:?]                             
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.evalInNewStackFrame(AbstractAnnotatedTree.java:293) ~[protelis-interpreter-13.0.3.jar:?]
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.lambda$projectAndEval$2(AbstractAnnotatedTree.java:248) ~[protelis-interpreter-13.0.3.jar:?]          
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.forEachWithIndex(AbstractAnnotatedTree.java:168) ~[protelis-interpreter-13.0.3.jar:?]
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.projectAndEval(AbstractAnnotatedTree.java:247) ~[protelis-interpreter-13.0.3.jar:?]                   
        at org.protelis.lang.interpreter.impl.AssignmentOp.evaluate(AssignmentOp.java:45) ~[protelis-interpreter-13.0.3.jar:?]
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:123) ~[protelis-interpreter-13.0.3.jar:?]                             
        at org.protelis.lang.interpreter.impl.All.lambda$evaluate$0(All.java:52) ~[protelis-interpreter-13.0.3.jar:?]                                       
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.forEachWithIndex(AbstractAnnotatedTree.java:168) ~[protelis-interpreter-13.0.3.jar:?]                 
        at org.protelis.lang.interpreter.impl.All.evaluate(All.java:50) ~[protelis-interpreter-13.0.3.jar:?]      
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:123) ~[protelis-interpreter-13.0.3.jar:?]                             
        at org.protelis.vm.impl.SimpleProgramImpl.compute(SimpleProgramImpl.java:65) ~[protelis-interpreter-13.0.3.jar:?]     
        at org.protelis.vm.ProtelisVM.runCycle(ProtelisVM.java:44) ~[protelis-interpreter-13.0.3.jar:?]                                                                   
        at com.bbn.protelis.networkresourcemanagement.NetworkServer.run(NetworkServer.java:355) ~[P2Protelis-8.2.3.jar:?]
        at com.bbn.protelis.networkresourcemanagement.NetworkServer.lambda$startExecuting$1(NetworkServer.java:415) ~[P2Protelis-8.2.3.jar:?]                             
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_222]                                                                                              
    Caused by: java.lang.NullPointerException: All does not allow null return values. In: true                                                                            
        at java.util.Objects.requireNonNull(Objects.java:290) ~[?:1.8.0_222]                                                                             
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.setAnnotation(AbstractAnnotatedTree.java:265) ~[protelis-interpreter-13.0.3.jar:?]                   
        at org.protelis.lang.interpreter.impl.ShareCall.lambda$evaluate$6(ShareCall.java:134) ~[protelis-interpreter-13.0.3.jar:?]                                       
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.forEachWithIndex(AbstractAnnotatedTree.java:168) ~[protelis-interpreter-13.0.3.jar:?]                
        at org.protelis.lang.interpreter.impl.ShareCall.evaluate(ShareCall.java:131) ~[protelis-interpreter-13.0.3.jar:?]
        at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:123) ~[protelis-interpreter-13.0.3.jar:?]
        ... 37 more
```

**Comment by DanySK** [comment]
#272 should work this around

**Comment by jschewebbn** [comment]
Do you have any ideas why Protelis 13 is unhappy with this file?

**Comment by jakebeal** [comment]
I think that there may be a bug introduced in the new version of the parser --- @DanySK , this isn't a stacktrace issue, it's an unexpected "required return values" issue.

**Comment by DanySK** [comment]
Hi Jon, the one in the first post? Does the failure happen at runtime or a compile time? If it's runtime (namely, if the stacktrace you attached is the one of the error you get), then the issue is that at some point you have a method call that returns `null`. Null values are strictly checked in Protelis, and their use must be wrapped in an `Option`.

```
methodCallReturningNull() // fails here
1
```
```
import org.protelis.lang.datatype.Option.*
fromNullable(methodCallReturningNull()) // fails here
1
```

**Comment by DanySK** [comment]
@jakebeal then I'm a bit lost.
Is the error: `Caused by: java.lang.NullPointerException: All does not allow null return values. In: true`? If not, maybe it would be a good idea to separate the conversations.

If so, then it's not an issue with the parser, as the parsing completes correctly, and there is a runtime failure during interpretation (hence a ProtelisRuntimeException is raised). Also, you can see that `runCycle` appears in the stacktrace, indicating a runtime failure.

**Comment by jakebeal** [comment]
Apologies, I meant interpreter rather than parser.  Nothing in here should be returning a null except for the "apDebugMessage" calls, which have a void return type.  These are never at the end of a block, however.

**Comment by DanySK** [comment]
Regardless, we currently have strict runtime checking for `null`. A method returning void returns `null` when queried for a result, and thus must be enclosed in an `Option`. Protelis at this stage just does not allow values to be `null`, not even as non-last call.
We may want to change this behavior, but that is what happens right now.

**Comment by jakebeal** [comment]
OK, that explains it then: that's a major backward non-compatible change that needs to be documented in the 13.0 release notes.  This will, indeed, need some sort of change in the future because otherwise it makes it difficult to access the (large) number of Java methods that return null.

For now, though: @jschewebbn, I think the patch is to make the debug calls return something --- maybe even just the string they create?

**Comment by jschewebbn** [comment]
@DanySK the function doesn't return null, it's void. So if I have in Java:
```
void foo(String arg) {
// do something
}
```
And then in Protelis the last line of a block is to call the Java function foo, I get error at runtime. My workaround is currently to change my Protelis from
```
{
...
foo("string")
}
```
to
```
{
...
foo("string")
0
}
```

It's not just the debug calls, there are all kinds of Java functions that we call at the end of a block that are void returns. A fairly common pattern is to call a set method at the end of a Protelis block with the data that has most recently been collected. I would prefer to just add a "0" to the end of each Protelis block that isn't expected to return anything.

**Comment by jschewebbn** [comment]
Note that the initial point of this ticket is to fix the not so useful error message when this crashes. Hopefully there is a release soon that includes #272 and I can figure out what line of my code is broken.

**Comment by DanySK** [comment]
The idea discussed in #183, and currently applied, is to do what most functional languages do, and provide an `Option` type which deals with computations that may not yield a value.
In your case, you should import `Option` (I can't recall if it's automatically imported) and switch your current:
```
{
...
foo("string")
0
}
```
to:
```
{
...
fromNullable(foo("string"))
}
```
This will make the function return an empty `Option`.

**Comment by jschewebbn** [comment]
Why convert void functions to returning null rather than returning an Optional like this? Then users wouldn't need to change code for their code to continue working?

**Comment by DanySK** [comment]
Because from the Protelis' point of view, void and null are the same thing. Remember: Protelis is by its nature functional, not imperative or OOP. The way functional languages deal with something similar to procedures is by having them returning a special type usually named [`Unit`](https://en.wikipedia.org/wiki/Unit_type).
Binding to an empty `Option` automatically is not a good workaround, as it assumes that the computation may not be empty. Solving this coherently would mean to either find a unit type within Java (`Void` is a class in Java, but I don't think it exposes any usable value besides the class itself), or to define our own unit type.
Probably the latter would be the cleanest. I'll open an issue about that.

**Comment by jschewebbn** [comment]
I'm not getting the error anymore with the development version. So I'm not sure if there is more to do on this ticket.

**Comment by DanySK** [comment]
Fixed in [13.1.0](https://github.com/Protelis/Protelis/releases/tag/13.1.0)
